### PR TITLE
new-log-viewer: Resize panel width when the browser window resizes.

### PR DESCRIPTION
**Pull request by junhaoliao**
# References

new-log-viewer series: #45 #46 #48 #51 #52 #53 #54 #55 #56 #59 #60 #61 #62 #63 #66 #67 #68 #69 #70 #71 #72 #73 #74 #76 #77 #78 #79 #80 #81 #82 #83 #84 #89 #91 #93

# Description

1. Refactor panel width setter / getters to use `document.documentElement`.
2. Call `getComputedStyle(document.documentElement)` instead of accessing `document.documentElement.style` directly.
3. Enforce a min width on the editor in place of the previous max panel width ratio approach.
4. Add window resize event handler to resize panel width when necessary.

# Validation performed

1. Started debug server as specified in #46.
2. Launch the app in a maximized browser window (Zoom: 100%) on a display larger than 1920x1080 (Scale: 1). Observed the panel was expanded by default.
3. Used the resize handle to adjust the width of the panel by dragging the handle towards the right of the window. Observed that the resizes were successful and that the panel cannot be further expanded when the editor's width becomes too small.
4. "Restore"d the browser window so it was no longer maximized. Reduced the window's width by dragging the resize handle towards the left and observed that as the panel reduced its width at the same time, while the editor remained visible with a fixed width. As the window becomes too small, the viewport could no longer hold both the editor and the panel, and the panel's width stopped reducing as the window's width reduced. Kept dragging the handle towards the left and the editor eventually became invisible.

## Summary by CodeRabbit

- **New Features**
	- Enhanced sidebar resizing logic to maintain a minimum width for the editor.
	- Dynamic adjustment of sidebar width during window resizing for improved user experience.

- **Bug Fixes**
	- Resolved issues with sidebar width calculations that could lead to improper display.

### CODE REVIEW

File: new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx [hunk 1]
```diff
@@ -82,16 +83,39 @@
...
+            }
+        };
+
+        window.addEventListener("resize", handleWindowResize);
+
+        return () => {
+            window.removeEventListener("resize", handleWindowResize);
+        };
+    }, []);
+
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx:118]: `getPanelWidth` is declared outside of the functional component and does not have any dependency on React states;
`tabListRef` is a ref which I believe does not need to be included in the dependency array.
`PANEL_CLIP_THRESHOLD_IN_PIXELS` is a const declared outside of the functional component, which does not change at all as well.

Do you think the dependency array is correctly specified now?

**Review by davemarco — CHANGES_REQUESTED** [review]
see earlier comment

**Comment by junhaoliao** [comment]
> If you click on the tab when the editor is small, the tab appears to "jump" to smaller length. I didn't investigate, but I feel it has something do to with this line `const availableWidth = window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS`
> 
>  Screen.Recording.2024-10-20.at.12.37.42.PM.mov

Should be now addressed with the tweaks in `PANEL_CLIP_THRESHOLD_IN_PIXELS` and `EDITOR_MIN_WIDTH_IN_PIXELS`.

**Comment by junhaoliao** [comment]
> If you click on the tab when the editor is small, the tab appears to "jump" to smaller length. I didn't investigate, but I feel it has something do to with this line `const availableWidth = window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS`
> 
>  Screen.Recording.2024-10-20.at.12.37.42.PM.mov

Should be now addressed with the tweaks in `PANEL_CLIP_THRESHOLD_IN_PIXELS` and `EDITOR_MIN_WIDTH_IN_PIXELS`.

**Comment by Henry8192** [comment]
> If you click on the tab when the editor is small, the tab appears to "jump" to smaller length. I didn't investigate, but I feel it has something do to with this line `const availableWidth = window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS`
> 
>  Screen.Recording.2024-10-20.at.12.37.42.PM.mov

I think I found the root cause: We need to also need to address `sidebar-tab-list-spacing` when we do the calculation. Though it is currently using flex-grow, I found the default value to be 48px for most of the case, so a hard code version is to deduct 48 whenever you do the caluculation.

For example, 
```ts
if (
    tabListRef.current.clientWidth + PANEL_CLIP_THRESHOLD_IN_PIXELS - 48 >
    resizeHandlePosition
) {
    // If the resize handle is positioned to the right of the <TabList/>'s right edge
    // with a clipping threshold accounted, close the panel.
    setPanelWidth(tabListRef.current.clientWidth);
}
```

**Comment by junhaoliao** [comment]
> > If you click on the tab when the editor is small, the tab appears to "jump" to smaller length. I didn't investigate, but I feel it has something do to with this line `const availableWidth = window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS`
> > Screen.Recording.2024-10-20.at.12.37.42.PM.mov
> 
> @junhaoliao I think I found the root cause: We need to also need to address `sidebar-tab-list-spacing` when we do the calculation. Though it is currently using flex-grow, I found the default value to be 48px for most of the case, so a hard code solution is to deduct 48 whenever you do the caluculation.
> 
> For example,
> 
> ```ts
> if (
>     tabListRef.current.clientWidth + PANEL_CLIP_THRESHOLD_IN_PIXELS - 48 >
>     resizeHandlePosition
> ) {
>     // If the resize handle is positioned to the right of the <TabList/>'s right edge
>     // with a clipping threshold accounted, close the panel.
>     setPanelWidth(tabListRef.current.clientWidth);
> }
> ```

I think your analysis is correct on that we did not account the sidebar width when doing the calculations. (We can confirm by double checking whether `tabListRef.current.clientWidth` is 48.)
Essentially, I believe we should not include `tabListRef.current.clientWidth` when comparing the `resizeHandlePosition` with the panel threshold. i.e., we can simply write:
```
        if (PANEL_CLIP_THRESHOLD_IN_PIXELS > resizeHandlePosition) {
            // If the resize handle is positioned to the right of the <TabList/>'s right edge
            // with a clipping threshold accounted, close the panel.
            setPanelWidth(tabListRef.current.clientWidth);
        }
...
```

**Review by davemarco — COMMENTED** [review]
I played around with current version, and found another bug. Here is my suggestions to fix issue. I think it works... Basically the idea is that you never let the window resize callback to resize smaller than the default panel. I also removed issue where resize handle was trying to set panel to a weird "available width"(when the window is very small)

File: new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx [hunk 2]
```diff
@@ -83,23 +97,33 @@ const Sidebar = () => {
 
             return;
         }
-        if (
-            tabListRef.current.clientWidth + PANEL_CLIP_THRESHOLD_IN_PIXELS >
-            resizeHandlePosition
-        ) {
+        if (PANEL_CLIP_THRESHOLD_IN_PIXELS > resizeHandlePosition) {
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx:100]: Similar to the one in handleActiveTabNameChange, I feel we want to use `Math.min(newPanelWidth, window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS)`

File: new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx [hunk 3]
```diff
@@ -83,23 +97,33 @@ const Sidebar = () => {
...
+            }
+        };
+
+        window.addEventListener("resize", handleWindowResize);
+
+        return () => {
+            window.removeEventListener("resize", handleWindowResize);
+        };
+    }, []);
+
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx:126]: Sorry for making a typo during refactoring, here's the intended logic. What do you think?
```
        const handleWindowResize = () => {
            const availableWidth = Math.max(
                window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS,
                PANEL_CLIP_THRESHOLD_IN_PIXELS
            );

            if (getPanelWidth() > availableWidth) {
                setPanelWidth(availableWidth);
            }
        };
```

File: new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx:87]: > that line in handleActiveTabNameChange is dangerous, as can set below the clipping width

I agree. Let's at least fix it in a way that mobile users can read the content in the panel; we can optimize the experience in a later PR.

[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/CentralContainer/Sidebar/index.tsx:95]: right, we kept the max and also added a min:
```
        clamp(
            window.innerWidth - EDITOR_MIN_WIDTH_IN_PIXELS,
            PANEL_CLIP_THRESHOLD_IN_PIXELS,
            PANEL_DEFAULT_WIDTH_IN_PIXELS
        )
```

FYR
```
/**
 * Clamps a number to the given range. E.g. If the number is greater than the range's upper bound,
 * the range's upper bound is returned.
 *
 * @param num The number to be clamped.
 * @param min The lower boundary of the output range.
 * @param max The upper boundary of the output range.
 * @return The clamped number.
 */
const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num, min), max);
```

**Comment by junhaoliao** [comment]
Latest demo:
https://github.com/user-attachments/assets/64a2eb1e-aa2b-495a-a3bb-8faffd286973

**Review by davemarco — APPROVED** [review]
I conditionally approve code. Functionality works as expected. Logic is not intuitive, but it works for now. We will have to redesign widths at a later date

**Pull request merged**